extension/json/json_deserializer.cpp

```diff
@@ -0,0 +1,306 @@
+#include "json_deserializer.hpp"
+#include "duckdb/common/types/blob.hpp"
+
+namespace duckdb {
+
+void JsonDeserializer::SetTag(const char *tag) {
+	current_tag = tag;
+}
+
+// If inside an object, return the value associated by the current tag (property name)
+// If inside an array, return the next element in the sequence
+yyjson_val *JsonDeserializer::GetNextValue() {
+	auto &parent_val = Current();
+	yyjson_val *val;
+	if (yyjson_is_obj(parent_val.val)) {
+		val = yyjson_obj_get(parent_val.val, current_tag);
+		if (!val) {
+			const char *json = yyjson_val_write(Current().val, 0, nullptr);
+			auto msg =
+			    StringUtil::Format("Expected but did not find property '%s' in json object: '%s'", current_tag, json);
+			free((void *)json);
+			throw ParserException(msg);
+		}
+	} else if (yyjson_is_arr(parent_val.val)) {
+		val = yyjson_arr_iter_next(&parent_val.arr_iter);
+		if (!val) {
+			const char *json = yyjson_val_write(Current().val, 0, nullptr);
+			auto msg =
+			    StringUtil::Format("Expected but did not find another value after exhausting json array: '%s'", json);
+			free((void *)json);
+			throw ParserException(msg);
+		}
+	} else {
+		// unreachable?
+		throw InternalException("Cannot get value from non-array/object");
+	}
+	return val;
+}
+
+void JsonDeserializer::ThrowTypeError(yyjson_val *val, const char *expected) {
+	auto actual = yyjson_get_type_desc(val);
+	auto &parent = Current();
+	if (yyjson_is_obj(parent.val)) {
```

Comment:
These two branches don't throw anything, and the string msg is discarded right away. Is that a bug?

Reply:
Yes, nice catch! I've pushed a fix in #8291